File: cpp/3_longest_substring_without_repeating_characters.cpp
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> letter_idx;
    int start = 0, max_len = 0;
    for (int end = 0; end < s.size(); end++) {
      char c = s[end];
      if (letter_idx.count(c) && letter_idx.at(c) >= start) {
        start = letter_idx.at(c) + 1;
      }
      letter_idx[c] = end;
      max_len = max(max_len, end - start + 1);
    }

    return max_len;
  }
};

File: rust/3_longest_substring_without_repeating_characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut letter_idx = std::collections::HashMap::new();
        let (mut start, mut max_len) = (0, 0);

        for (end, c) in s.as_bytes().iter().enumerate() {
            if letter_idx.contains_key(c) && letter_idx[c] >= start {
                start = letter_idx[c] + 1;
            }
            letter_idx.insert(c, end);
            max_len = std::cmp::max(max_len, end - start + 1);
        }

        max_len as i32
    }
}
